File: bindings/java/Omid.java
package omid;

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Pointer;
import com.sun.jna.Structure;

public interface Omid extends Library {
    Omid INSTANCE = Native.load("omid", Omid.class);

    @Structure.FieldOrder({"objectId", "eventType", "flags", "payload"})
    class OmidPacket extends Structure {
        public short objectId;
        public byte eventType;
        public byte flags;
        public int payload;

        public static class ByValue extends OmidPacket implements Structure.ByValue {}
    }

    @Structure.FieldOrder({"objectId", "objectType", "spatialX", "spatialY", "resolution"})
    class TopologyDescriptor extends Structure {
        public short objectId;
        public byte objectType;
        public short spatialX;
        public short spatialY;
        public byte resolution;

        public static class ByValue extends TopologyDescriptor implements Structure.ByValue {}
    }

    OmidPacket.ByValue omid_packet_new(short objectId, byte eventType, byte flags, int payload);
    OmidPacket.ByValue omid_packet_new_f32(short objectId, byte eventType, byte flags, float val);
    OmidPacket.ByValue omid_packet_new_i32(short objectId, byte eventType, byte flags, int val);
    OmidPacket.ByValue omid_packet_new_u32(short objectId, byte eventType, byte flags, int val);
    OmidPacket.ByValue omid_packet_new_xy(short objectId, byte eventType, byte flags, short x, short y);
    OmidPacket.ByValue omid_packet_new_haptic(short objectId, byte profile, float intensity);
    OmidPacket.ByValue omid_packet_new_adc12(short objectId, byte eventType, byte flags, short val);
    OmidPacket.ByValue omid_packet_new_adc16(short objectId, byte eventType, byte flags, short val);
    OmidPacket.ByValue omid_packet_from_bytes(Pointer bytes);
    void omid_packet_to_bytes(OmidPacket.ByValue packet, Pointer out_bytes);
    float omid_packet_payload_as_f32(OmidPacket.ByValue packet);
    int omid_packet_payload_as_i32(OmidPacket.ByValue packet);
    int omid_packet_payload_as_u32(OmidPacket.ByValue packet);
    void omid_packet_payload_as_xy(OmidPacket.ByValue packet, Pointer out_x, Pointer out_y);
    short omid_packet_payload_as_adc12(OmidPacket.ByValue packet);
    short omid_packet_payload_as_adc16(OmidPacket.ByValue packet);
    float omid_packet_payload_as_normalized_f32(OmidPacket.ByValue packet, byte adc_bits);
    float omid_packet_haptic_intensity(OmidPacket.ByValue packet);
    int omid_packet_haptic_force_profile(OmidPacket.ByValue packet);

    byte omid_flags_new(boolean touched, boolean raw_data, boolean direction, byte timer_delta);
    boolean omid_flags_is_touched(byte flags);
    boolean omid_flags_is_raw_data(byte flags);
    boolean omid_flags_direction(byte flags);
    byte omid_flags_timer_delta(byte flags);

    TopologyDescriptor.ByValue omid_topology_new(short objectId, byte objectType, short spatialX, short spatialY, byte resolution);
    TopologyDescriptor.ByValue omid_topology_from_bytes(Pointer bytes);
    void omid_topology_to_bytes(TopologyDescriptor.ByValue desc, Pointer out_bytes);
}

File: bindings/kotlin/Omid.kt
package omid

import com.sun.jna.Pointer
import com.sun.jna.Memory

enum class EventType(val value: Byte) {
    Unknown(0x00),
    AbsoluteChange(0x01),
    RelativeDelta(0x02),
    KeyPress(0x03),
    KeyRelease(0x04),
    HapticFeedback(0x05),
    VisualUpdate(0x06),
    SystemHandshake(0x07);

    companion object {
        fun fromByte(value: Byte): EventType =
            values().find { it.value == value } ?: Unknown
    }
}

enum class ForceProfile(val value: Byte) {
    Unknown(0x00),
    HammerStrike(0x01),
    SpringTension(0x02),
    KineticDampening(0x03);

    companion object {
        fun fromByte(value: Byte): ForceProfile =
            values().find { it.value == value } ?: Unknown
    }
}

class OmidFlags(val raw: Byte) {
    val isTouched: Boolean get() = Omid.INSTANCE.omid_flags_is_touched(raw)
    val isRawData: Boolean get() = Omid.INSTANCE.omid_flags_is_raw_data(raw)
    val direction: Boolean get() = Omid.INSTANCE.omid_flags_direction(raw)
    val timerDelta: Byte get() = Omid.INSTANCE.omid_flags_timer_delta(raw)

    companion object {
        fun create(touched: Boolean, rawData: Boolean, direction: Boolean, timerDelta: Byte): OmidFlags {
            val raw = Omid.INSTANCE.omid_flags_new(touched, rawData, direction, timerDelta)
            return OmidFlags(raw)
        }
    }
}

class Packet(val inner: Omid.OmidPacket.ByValue) {
    val objectId: Short get() = inner.objectId
    val eventType: EventType get() = EventType.fromByte(inner.eventType)
    val flags: OmidFlags get() = OmidFlags(inner.flags)
    val payload: Int get() = inner.payload

    fun payloadAsF32(): Float = Omid.INSTANCE.omid_packet_payload_as_f32(inner)
    fun payloadAsI32(): Int = Omid.INSTANCE.omid_packet_payload_as_i32(inner)
    fun payloadAsU32(): Int = Omid.INSTANCE.omid_packet_payload_as_u32(inner)
    
    fun payloadAsXY(): Pair<Short, Short> {
        val outX = Memory(2)
        val outY = Memory(2)
        Omid.INSTANCE.omid_packet_payload_as_xy(inner, outX, outY)
        return Pair(outX.getShort(0), outY.getShort(0))
    }

    fun payloadAsAdc12(): Short = Omid.INSTANCE.omid_packet_payload_as_adc12(inner)
    fun payloadAsAdc16(): Short = Omid.INSTANCE.omid_packet_payload_as_adc16(inner)
    fun payloadAsNormalizedF32(adcBits: Byte): Float = Omid.INSTANCE.omid_packet_payload_as_normalized_f32(inner, adcBits)
    
    fun hapticIntensity(): Float = Omid.INSTANCE.omid_packet_haptic_intensity(inner)
    fun hapticForceProfile(): ForceProfile {
        val profileVal = Omid.INSTANCE.omid_packet_haptic_force_profile(inner)
        return if (profileVal >= 0) ForceProfile.fromByte(profileVal.toByte()) else ForceProfile.Unknown
    }

    fun toBytes(): ByteArray {
        val mem = Memory(8)
        Omid.INSTANCE.omid_packet_to_bytes(inner, mem)
        return mem.getByteArray(0, 8)
    }

    companion object {
        fun create(objectId: Short, eventType: EventType, flags: OmidFlags, payload: Int): Packet =
            Packet(Omid.INSTANCE.omid_packet_new(objectId, eventType.value, flags.raw, payload))

        fun createF32(objectId: Short, eventType: EventType, flags: OmidFlags, value: Float): Packet =
            Packet(Omid.INSTANCE.omid_packet_new_f32(objectId, eventType.value, flags.raw, value))

        fun createI32(objectId: Short, eventType: EventType, flags: OmidFlags, value: Int): Packet =
            Packet(Omid.INSTANCE.omid_packet_new_i32(objectId, eventType.value, flags.raw, value))

        fun createXY(objectId: Short, eventType: EventType, flags: OmidFlags, x: Short, y: Short): Packet =
            Packet(Omid.INSTANCE.omid_packet_new_xy(objectId, eventType.value, flags.raw, x, y))

        fun createHaptic(objectId: Short, profile: ForceProfile, intensity: Float): Packet =
            Packet(Omid.INSTANCE.omid_packet_new_haptic(objectId, profile.value, intensity))

        fun createAdc12(objectId: Short, eventType: EventType, flags: OmidFlags, value: Short): Packet =
            Packet(Omid.INSTANCE.omid_packet_new_adc12(objectId, eventType.value, flags.raw, value))

        fun createAdc16(objectId: Short, eventType: EventType, flags: OmidFlags, value: Short): Packet =
            Packet(Omid.INSTANCE.omid_packet_new_adc16(objectId, eventType.value, flags.raw, value))

        fun fromBytes(bytes: ByteArray): Packet {
            require(bytes.size == 8) { "Data must be exactly 8 bytes" }
            val mem = Memory(8)
            mem.write(0, bytes, 0, 8)
            return Packet(Omid.INSTANCE.omid_packet_from_bytes(mem))
        }
    }
}
